tests/unit/test_runtime_build.py

```diff
@@ -290,3 +309,234 @@ def test_build_runtime_image_exact_hash_not_exist(mock_build_sandbox_image, temp
+                f'{unique_prefix}:base',
+                f'{unique_prefix}:intermediate',
+            ]
+
+            print(f'\nImage ID: {image.id}')
+            print(f'Image tags: {all_tags}\n')
+
+            yield image
+
+        finally:
+            # Clean up all tagged images
+            for tag in all_tags:
+                try:
+                    client.images.remove(tag, force=True)
+                    print(f'Removed image: {tag}')
+                except Exception as e:
+                    print(f'Error removing image {tag}: {str(e)}')
+
+
+def test_init(docker_runtime_builder):
+    assert isinstance(docker_runtime_builder.docker_client, docker.DockerClient)
+    assert docker_runtime_builder.max_lines == 10
+    assert docker_runtime_builder.log_lines == [''] * 10
+
+
+def test_build_image_from_scratch(docker_runtime_builder, tmp_path):
+    context_path = str(tmp_path)
+    tags = ['test_build:latest']
+
+    # Create a minimal Dockerfile in the context path
+    with open(os.path.join(context_path, 'Dockerfile'), 'w') as f:
+        f.write("""FROM php:latest
+CMD ["sh", "-c", "echo 'Hello, World!'"]
+""")
+    built_image_name = None
+    container = None
+    client = docker.from_env()
+    try:
+        with patch('sys.stdout.isatty', return_value=False):
+            built_image_name = docker_runtime_builder.build(
+                context_path,
+                tags,
+                use_local_cache=False,
+            )
+            assert built_image_name == f'{tags[0]}'
+
+            # Verify the image was created
+            image = client.images.get(tags[0])
+            assert image is not None
+
+    except docker.errors.ImageNotFound:
+        pytest.fail('test_build_image_from_scratch: test image not found!')
+    except Exception as e:
+        pytest.fail(f'test_build_image_from_scratch: Build failed with error: {str(e)}')
+
+    finally:
+        # Clean up the container
+        if container:
+            try:
+                container.remove(force=True)
+                logger.info(f'Removed test container: `{container.id}`')
+            except Exception as e:
+                logger.warning(
+                    f'Failed to remove test container `{container.id}`: {str(e)}'
+                )
+
+        # Clean up the image
+        if built_image_name:
+            try:
+                client.images.remove(built_image_name, force=True)
+                logger.info(f'Removed test image: `{built_image_name}`')
+            except Exception as e:
+                logger.warning(
+                    f'Failed to remove test image `{built_image_name}`: {str(e)}'
+                )
+        else:
+            logger.warning('No image was built, so no image cleanup was necessary.')
+
+
+def _format_size_to_gb(bytes_size):
+    """Convert bytes to gigabytes with two decimal places."""
+    return round(bytes_size / (1024**3), 2)
+
+
+def test_list_dangling_images():
```

Comment:
Is this an actual test that need to assert stuff, or just a utility function to make sure things goes smoothly?

Reply:
Sortof both as it uses the "filters" arg. And I was curious what it would show on my local WSL. :)